Don't request line_items in the order list

joshbetz opened the pull request:
In the order list view, we don't show details about the individual
products in the order. Requesting full details about all the products
can have a significant impact on performance.

Here we differentiate between fields we need for single orders and a
list of orders. For now the only difference is that single orders do
load the line items.

joshbetz committed:
Don't request line_items in the order list

In the order list view, we don't show details about the individual
products in the order. Requesting full details about all the products
can have a significant impact on performance.

Here we differentiate between fields we need for single orders and a
list of orders. For now the only difference is that single orders do
load the line items.

Ecarrion commented on Networking/Networking/Model/Order.swift:
```diff
@@ -133,7 +133,7 @@ public struct Order: Decodable, GeneratedCopiable, GeneratedFakeable {
         let paymentMethodTitle = try container.decode(String.self, forKey: .paymentMethodTitle)
 
-        let items = try container.decode([OrderItem].self, forKey: .items)
+        let items = try? container.decode([OrderItem].self, forKey: .items)
```
To avoid modifying the initializer and to not hide/shadow a possible parsing error, you could do 

```suggestion
        let items = try? container.decodeIfPresent([OrderItem].self, forKey: .items) ?? []
```

Ecarrion commented on Networking/Networking/Remote/OrdersRemote.swift:
```diff
@@ -199,11 +199,16 @@ public extension OrdersRemote {
 
     enum ParameterValues {
-        static let fieldValues: String = """
+        static let listFieldValues: String = """
```
We should probably add a comment telling the reader that the only field missing is `line_items` to make this easy to understand in the future!

Ecarrion reviewed (approved):
Congrats on your first? PR!

Changes look good, I only left some minor suggestions!

Testing worked well but I didn't find any improvements on the request loading times for my test store, both(this branch and develop) were in the `0.8s~1s` ranges, not sure if there is some cache going around tho.

The experience is a bit less ideal since now there is another section that needs to be filled and jumps tho.

https://user-images.githubusercontent.com/562080/133508244-686afcbd-cb10-4a63-8f02-acaf843e4e8e.mov

What loading times are you seeing @joshbetz?

joshbetz committed:
avoid modifying the initializer and to not hide/shadow a possible parsing error

props @ecarrion

Co-authored-by: Ernesto Carrión <ernesto.carrion@automattic.com>

joshbetz commented:
For each test (with and without line_items), I made two requests in a row. The second request was a bit faster each time, but regardless, there is a clear improvement (~33% in this case) by removing line_items when it's not needed on this test site.

With line_items:

```
$ curl -w 'ttfb: %{time_starttransfer}\ntotal: %{time_total}' -so /dev/null -u '<redacted>:<redacted>' 'https://teststore.betzster.com/wp-json/wc/v3/orders?_fields=id,parent_id,number,status,currency,customer_id,customer_note,date_created_gmt,date_modified_gmt,date_paid_gmt,discount_total,discount_tax,shipping_total,shipping_tax,total,total_tax,payment_method,payment_method_title,shipping,billing,coupon_lines,shipping_lines,refunds,fee_lines,line_items'
ttfb: 1.254984
total: 1.299505

$ curl -w 'ttfb: %{time_starttransfer}\ntotal: %{time_total}' -so /dev/null -u '<redacted>:<redacted>' 'https://teststore.betzster.com/wp-json/wc/v3/orders?_fields=id,parent_id,number,status,currency,customer_id,customer_note,date_created_gmt,date_modified_gmt,date_paid_gmt,discount_total,discount_tax,shipping_total,shipping_tax,total,total_tax,payment_method,payment_method_title,shipping,billing,coupon_lines,shipping_lines,refunds,fee_lines,line_items'
ttfb: 0.927129
total: 0.971201
```

Without line_items:

```
$ curl -w 'ttfb: %{time_starttransfer}\ntotal: %{time_total}' -so /dev/null -u '<redacted>:<redacted>' 'https://teststore.betzster.com/wp-json/wc/v3/orders?_fields=id,parent_id,number,status,currency,customer_id,customer_note,date_created_gmt,date_modified_gmt,date_paid_gmt,discount_total,discount_tax,shipping_total,shipping_tax,total,total_tax,payment_method,payment_method_title,shipping,billing,coupon_lines,shipping_lines,refunds,fee_lines'
ttfb: 0.874923
total: 0.875409

$ curl -w 'ttfb: %{time_starttransfer}\ntotal: %{time_total}' -so /dev/null -u '<redacted>:<redacted>' 'https://teststore.betzster.com/wp-json/wc/v3/orders?_fields=id,parent_id,number,status,currency,customer_id,customer_note,date_created_gmt,date_modified_gmt,date_paid_gmt,discount_total,discount_tax,shipping_total,shipping_tax,total,total_tax,payment_method,payment_method_title,shipping,billing,coupon_lines,shipping_lines,refunds,fee_lines'
ttfb: 0.614476
total: 0.617481
```

joshbetz commented:
One issue that should definitely get fixed though, is that we see the "Loading content" and jump even if you go back and forth between the list of orders and a given order multiple times. Should that get cached locally after we initially load an order?


https://user-images.githubusercontent.com/300429/133514615-b5c51a41-6905-46fd-86bc-c0cd6319ed59.mp4

joshbetz committed:
Remove billing and shipping from order list request

Billing and shipping information can have a big impact on the size of
the response and are also not needed in the order list.

joshbetz committed:
Adding billing info back

We use this on the order list screen to show the show the name of the
person on the order.

joshbetz committed:
Prevent order items from updating if items is empty

When we go from the single order screen to the list of orders, we update
the order list. Since the oder list request no longer includes line
items, updating the order items ends up deleting them all. Then when we
go back to the single order screen, it has to fill in the list again.
Going back and forth between the order list and a given order will cause
it to have to fill in the list of orders every time.

If the order items is empty, it's a pretty good sign that we're on the
list. However, it would be better to allow the order items to be nil and
then we could look for `readOnlyOrder.items != nil`, which would be a
better indication that the data just doesn't include line items.

joshbetz commented:
@Ecarrion 21bb361a6519e33e008b3a3fbd1a6b2618dec174 "fixes" the issue I mentioned above. I think it would be better to let `items` be `nil` to differentiate between an empty array and the field not existing. What are your thoughts?

Ecarrion commented:
@joshbetz 

>  I think it would be better to let items be nil to differentiate between an empty array and the field not existing. What are your thoughts?

I think that makes sense, then, we should probably do the same for `shipping` that I saw you are also removing!

joshbetz commented:
shipping address is already optional https://github.com/woocommerce/woocommerce-ios/blob/develop/Networking/Networking/Model/Order.swift#L36

joshbetz commented:
@Ecarrion Any suggestions on the best way to make that optional? It seems like there are a lot of places that use `items` that will need to be fixed.

Ecarrion commented:
> @Ecarrion Any suggestions on the best way to make that optional? It seems like there are a lot of places that use `items` that will need to be fixed.

Hey @joshbetz, I don't think there is an easy and clean way to handle that update. 🤔 

_Maybe,_ we could create a [property wrapper](https://www.swiftbysundell.com/articles/property-wrappers-in-swift/) that returns an empty array if the array is `nil`, but still allows us to know if the underlying value is `nil` or not.

Since property wrappers are a controversial topic, I think the best way to move forward is to leave the code with your change, and add a comment on why we need that `items.count > 0` check(btw we can also use `items.isEmpty`.)

Then create an issue where we explain the issue and the property wrapper solution so the team can debate it!

What do you think?

joshbetz commented:
Ok cool, that seems reasonable to me.

Ecarrion commented:
Hey @joshbetz do you need something to get this merged?

joshbetz commented:
@Ecarrion If it looks good to you, I'm happy to merge.

Ecarrion commented:
@joshbetz Yup, I approved it a while ago 😅 

Maybe just adding the comment we discussed here https://github.com/woocommerce/woocommerce-ios/pull/4997/files#diff-07ad979afe48bc7bcfd986f5c452866c93cc06133b6e13aa34adcec76ac2c31eR71 and then merge away!

joshbetz committed:
Add comment to clarify the listFieldValues
